### PR TITLE
SMC Capability

### PR DESCRIPTION
**Pull request by Furao**
Add a new capability which certain threads can invoke so seL4 can make SMC calls on ARM platforms in EL2 (virtualized mode) on the thread’s behalf.

See https://sel4.atlassian.net/browse/RFC-9

### CODE REVIEW

File: libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml [hunk 1]
```diff
@@ -44,6 +44,16 @@
...
     </struct>
+    <struct name="seL4_ARM_SMCContext">
+        <member name="x0"/>
+        <member name="x1"/>
+        <member name="x2"/>
+        <member name="x3"/>
+        <member name="x4"/>
+        <member name="x5"/>
+        <member name="x6"/>
+        <member name="x7"/>
```
Comment (yyshen) [inline libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml:55]: might need to state that this only works for SMC calls that follow the SMC calling convention.

File: src/arch/arm/config.cmake [hunk 2]
```diff
@@ -194,6 +194,12 @@ config_option(
 )
 mark_as_advanced(KernelAArch64SErrorIgnore)
 
+config_option(
+    KernelAllowSMCCalls ALLOW_SMC_CALLS "Allow virtualized guests to make SMC calls"
+    DEFAULT OFF
```
Comment (yyshen) [inline src/arch/arm/config.cmake:199]: probably need to enable the option that traps SMC calls?

File: src/arch/arm/object/smc.c [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+exception_t decodeARMSMCInvocation(word_t label, unsigned int length, cptr_t cptr,
+                                  cte_t *srcSlot, cap_t cap, bool_t call, word_t *buffer)
+{
+    word_t badge;
+    word_t smc_func_id;
+
+    switch (label) {
+    case ARMSMCCall:
+        badge = cap_smc_cap_get_capSMCBadge(cap);
+        smc_func_id = getSyscallArg(0, buffer);
```
Comment (yyshen) [inline src/arch/arm/object/smc.c:69]: so the badge is used to represent the authorised smc_func_id for a minted SMC Cap. by default, the original  SMC Cap can authorise all smc func calls, right?

File: src/arch/arm/object/smc.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    register seL4_Word r0 asm("x0") = arg[0];
+    register seL4_Word r1 asm("x1") = arg[1];
+    register seL4_Word r2 asm("x2") = arg[2];
+    register seL4_Word r3 asm("x3") = arg[3];
+    register seL4_Word r4 asm("x4") = arg[4];
+    register seL4_Word r5 asm("x5") = arg[5];
+    register seL4_Word r6 asm("x6") = arg[6];
+    register seL4_Word r7 asm("x7") = arg[7];
+    asm volatile("smc #0\n"
```
Comment (yyshen) [inline src/arch/arm/object/smc.c:31]: have you considered to release the lock for the SMP kernel before the call and acquire the lock again after the call. The smc call can run arbitrary long.

File: libsel4/include/sel4/bootinfo_types.h [hunk 5]
```diff
@@ -24,11 +24,12 @@ enum {
     seL4_CapDomain              = 11, /* global domain controller cap */
     seL4_CapSMMUSIDControl      = 12,  /*global SMMU SID controller cap, null cap if not supported*/
     seL4_CapSMMUCBControl       = 13,  /*global SMMU CB controller cap, null cap if not supported*/
+    seL4_CapSMC                 = 14,  /*global SMC cap, null cap if not supported*/
```
Comment (yyshen) [inline libsel4/include/sel4/bootinfo_types.h:27]: It looks like one SMC cap is used for all CPU cores, not sure if this is coarse-grained.

File: src/arch/arm/object/smc.c (hunk 3 above)
Comment (lsf37) [reply src/arch/arm/object/smc.c:69]: According to the code, badge 0 authorises all, which would be the same convention as for endpoints. I think that makes sense. We should check though, if smc call 0 is reserved or may be needed, because otherwise anyone who needs to do that call can do all calls.

File: src/arch/arm/object/smc.c (hunk 4 above)
Comment (lsf37) [reply src/arch/arm/object/smc.c:31]: Hm, that is an important observation -- I hadn't thought about WCET on these yet at all, and without knowing more about individual calls, the kernel WCET for a system with SMC enabled is now arbitrarily high. That is even true on uni-core, but it is worse on SMP, because the kernel is blocked for all cores during that time.

This is different to other user-level behaviour that can take arbitrarily long, because this is now kernel time.

I don't think that releasing the lock while inside the kernel works in the current design: it wouldn't be a big-lock any more, and when you re-acquire the lock you would have not guarantee that the current TSC, SC, or IPC buffer still exists (some other core could have retyped them, for instance). I'm sure that could be worked-around, but it would definitely break the big-lock idea. And even if that should somehow work, it would still not solve the uni-core case.

That might throw a real spanner in the works. Do we need to turn this into some kind of more asynchronous mechanism?

File: libsel4/include/sel4/bootinfo_types.h (hunk 5 above)
Comment (lsf37) [reply libsel4/include/sel4/bootinfo_types.h:27]: That looks fine to me, as long as the intended pattern is that the system initialiser makes badged copies of this cap, handing these out to whoever needs them, restricted to core, sub-system or whichever is needed. Usually the 0-badge version of the cap should stay in the initialiser, or in a high-privilege management component.

That scenario does suggest that we might want more derivation levels for this cap, though, like for endpoint caps.

File: libsel4/arch_include/arm/interfaces/sel4arch.xml [hunk 6]
```diff
@@ -23,7 +23,7 @@
             <param dir="in" name="vaddr" type="seL4_Word"
             description="Virtual address to map the page into."/>
             <param dir="in" name="attr" type="seL4_ARM_VMAttributes">
-                <description> 
+                <description>
                     VM Attributes for the mapping. <docref>Possible values for this type are given in <autoref label="ch:vspace"/>  .</docref>
```
Comment (axel-h) [inline libsel4/arch_include/arm/interfaces/sel4arch.xml:27]: Could you move these space changes to a dedicated commit?

File: libsel4/include/sel4/bootinfo_types.h (hunk 5 above)
Comment (axel-h) [inline libsel4/include/sel4/bootinfo_types.h:27]: Is this a potentially breaking change that should be noted in the readme?

File: libsel4/arch_include/arm/interfaces/sel4arch.xml (hunk 6 above)
Comment (Furao) [reply libsel4/arch_include/arm/interfaces/sel4arch.xml:27]: Yeah, I can do that.

File: libsel4/include/sel4/bootinfo_types.h (hunk 5 above)
Comment (Furao) [reply libsel4/include/sel4/bootinfo_types.h:27]: That is a good point I hadn't considered. Adding to the end avoids breaking old layouts, but that becomes much harder to do with the sched_context cap getting configured out. Maybe positions should be constant whether their features are enabled or not? 
ex:
```
enum {
    seL4_CapNull                =  0, /* null cap */
    seL4_CapInitThreadTCB       =  1, /* initial thread's TCB cap */
    seL4_CapInitThreadCNode     =  2, /* initial thread's root CNode cap */
    seL4_CapInitThreadVSpace    =  3, /* initial thread's VSpace cap */
    seL4_CapIRQControl          =  4, /* global IRQ controller cap */
    seL4_CapASIDControl         =  5, /* global ASID controller cap */
    seL4_CapInitThreadASIDPool  =  6, /* initial thread's ASID pool cap */
    seL4_CapIOPortControl       =  7, /* global IO port control cap (null cap if not supported) */
    seL4_CapIOSpace             =  8, /* global IO space cap (null cap if no IOMMU support) */
    seL4_CapBootInfoFrame       =  9, /* bootinfo frame cap */
    seL4_CapInitThreadIPCBuffer = 10, /* initial thread's IPC buffer frame cap */
    seL4_CapDomain              = 11, /* global domain controller cap */
    seL4_CapSMMUSIDControl      = 12,  /*global SMMU SID controller cap, null cap if not supported*/
    seL4_CapSMMUCBControl       = 13,  /*global SMMU CB controller cap, null cap if not supported*/
    seL4_CapInitThreadSC        = 14, /* initial thread's scheduling context cap */
    seL4_CapSMC                    = 15,  /*global SMC cap, null cap if not supported*/
    seL4_NumInitialCaps         = 16
};
```

Then if they feature isn't enabled, it just shows up as a nullcap?

File: src/arch/arm/object/smc.c (hunk 4 above)
Comment (axel-h) [reply src/arch/arm/object/smc.c:31]: if the short term goal is, that the feature is to be merged (behind a config option) , I'm fine with the comment https://sel4.atlassian.net/browse/RFC-9?focusedCommentId=19762 from @lsf37 that for now the system designer needs to be aware of the potential pitfalls. We can look into this as a follow-up then.

Comment (Furao) [reply src/arch/arm/object/smc.c:31]: The warning message has been added here: https://github.com/seL4/seL4/pull/701/files#diff-bc9402a3a9f4e827bf5919980dd881d416c5a6225a82a380af72117a605ae299R198-R203

File: src/arch/arm/64/object/objecttype.c [hunk 7]
```diff
@@ -495,6 +501,10 @@ exception_t Arch_decodeInvocation(word_t label, word_t length, cptr_t cptr,
     case cap_cb_cap:
         return decodeARMCBInvocation(label, length, cptr, slot, cap, call, buffer);
 #endif /*CONFIG_ARM_SMMU*/
+#ifdef CONFIG_ALLOW_SMC_CALLS
```
Comment (axel-h) [inline src/arch/arm/64/object/objecttype.c:516]: I think we have to add this in line 488 also

Comment (Furao) [reply src/arch/arm/64/object/objecttype.c:516]: CONFIG_ALLOW_SMC_CALLS has a dependency on CONFIG_ARM_HYPERVISOR_SUPPORT, so it works how it is. However, we can be more explicit and add it into 488 if you would like.

Comment (axel-h) [reply src/arch/arm/64/object/objecttype.c:516]: I think having this explicitly improves maintainability. How strong us the dependency of the SMC feature on the Hypervisor feature - could they be separated one day?

Comment (Furao) [reply src/arch/arm/64/object/objecttype.c:516]: We have actually already seen a use case to uncouple this from virtualization support, so we updated the specified ifdef line and changed the config dependency.

**Comment by Furao** [comment]
@axel-h Ping

**Comment by axel-h** [comment]
Could you rebase this in the current master?

File: libsel4/include/sel4/bootinfo_types.h (hunk 5 above)
Comment (axel-h) [reply libsel4/include/sel4/bootinfo_types.h:27]: @lsf37, @kent-mcleod; What do you think about this? 

A change here might also require updating `SIMPLE_SKIPPED_INIT_CAPS` here:
- https://github.com/seL4/seL4_libs/blob/master/libsel4simple/arch_include/arm/simple/arch/simple.h#L17-L19 (note that there is no RISC-V here, it uses the ARM files)
- https://github.com/seL4/seL4_libs/blob/master/libsel4simple/arch_include/x86/simple/arch/simple.h#L17-L23

**Comment by Furao** [comment]
Ok, rebased.

File: libsel4/include/sel4/bootinfo_types.h (hunk 5 above)
Comment (Furao) [reply libsel4/include/sel4/bootinfo_types.h:28]: @lsf37, @kent-mcleod; ping.

Comment (Furao) [reply libsel4/include/sel4/bootinfo_types.h:28]: @lsf37, @kent-mcleod; pinging again.

File: src/arch/arm/config.cmake [hunk 8]
```diff
@@ -194,6 +194,17 @@ config_option(
...
 
+config_option(
+    KernelAllowSMCCalls ALLOW_SMC_CALLS "Allow components to make SMC calls.
+    WARNING: Some platforms require SMC calls to actually be made to boot VMs \
+    correctly, however allowing SMC calls causes a couple of issues. Since seL4 \
+    cannot pre-empt the supervisor the WCET is no longer guaranteed. Also, \
+    since the supervisor is a higher privilege level and can make any change \
+    in the system, the proofs can no longer be guaranteed."
+    DEFAULT OFF
+    DEPENDS "KernelSel4ArchAarch64"
```
[the remark this answers is not in the document]
Comment (Furao) [reply src/arch/arm/config.cmake:205]: Yeah, that sounds like a good idea. I will make that update.

File: libsel4/include/sel4/bootinfo_types.h (hunk 5 above)
Comment (lsf37) [reply libsel4/include/sel4/bootinfo_types.h:28]: Apologies for the long silence. As commented in #1088, from the verification side that would be Ok.

Comment (kent-mcleod) [reply libsel4/include/sel4/bootinfo_types.h:28]: Once this is rebased on the changes from #1088, this should be resolved?

**Comment by kent-mcleod** [comment]
@Furao, I added some suggested changes but can't PR off this PR due to a rebase.  https://github.com/seL4/seL4/compare/master...kent-mcleod:seL4:kent/smc.  Are you happy to pull these additional commits in?  In addition to some changes to the invocation implementation to more closely follow the structures required by the verification specs, I've made it so that seL4_CNode_Revoke should do the right thing:
- Revoking the root's original smc cap should delete all SMC caps in the system,
- Revoking the first badged smc cap should delete any SMC caps derived from it.

File: src/arch/arm/config.cmake (hunk 8 above)
[the remark this answers is not in the document]
Comment (Furao) [reply src/arch/arm/config.cmake]: Updated this dependency in the latest.

File: src/object/cnode.c [hunk 9]
```diff
@@ -806,6 +806,21 @@ bool_t PURE isMDBParentOf(cte_t *cte_a, cte_t *cte_b)
         break;
     }
 
+#ifdef CONFIG_ALLOW_SMC_CALLS
```
Comment (kent-mcleod) [inline src/object/cnode.c:809]: This would benefit from some arch-specific functions to call out to, but they don't currently exist... Thoughts?

File: src/object/cnode.c [hunk 10]
```diff
@@ -806,6 +806,21 @@ bool_t PURE isMDBParentOf(cte_t *cte_a, cte_t *cte_b)
...
+    case cap_smc_cap: {
+        word_t badge;
+
+        badge = cap_smc_cap_get_capSMCBadge(cte_a->cap);
+        if (badge == 0) {
+            return true;
+        }
+        return
```
Comment (Indanz) [inline src/object/cnode.c:817]: The formatting of this return is very weird. I know that cap_notification_cap does the same, but I don't think there is a reason to spread this further.

File: libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml [hunk 11]
```diff
@@ -207,4 +217,18 @@
             </error>
         </method>
     </interface>
+    <interface name="seL4_ARM_SMC" manual_name="SMC" cap_description="Capability to allow threads to interact with SMC calls.">
+        <method id="ARMSMCCall" name="Call" manual_name="SMC Call" manual_label="smc_call">
+            <brief>
+                Tell the microkernel to make the real SMC call.
+            </brief>
+            <description>
+                <docref>See <autoref label="sec:read_write_registers"/></docref>
```
Comment (Indanz) [inline libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml:226]: Is this reference still up-to-date, considering that `seL4_ARM_SMCContext` has been added?

File: src/arch/arm/object/smc.c [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+                 : "+r"(r0), "+r"(r1), "+r"(r2), "+r"(r3)
+                 : "r"(r4), "r"(r5), "r"(r6), "r"(r7)
```
Comment (Indanz) [inline src/arch/arm/object/smc.c:32]: According to DEN0028E_SMC_Calling_Convention_1.4eac0:

- A Function Identifier is passed in register W0.
- Arguments are passed in registers W1-W7.
- Results are returned in W0-W7.
- Registers W4-W7 must be preserved unless they contain results, as specified in the function definition.

(Same for 64-bit calls, but with X0-X17 instead.)

So I think both the output and input lists should be extended to the r0-r7 registers, both with `+r`.

Also, if only the 32-bit calling convention is supported, the code should check that bit 30 is not set in `smc_func_id`.

Comment (Furao) [reply src/arch/arm/object/smc.c:32]: So we are actually targeting a specific version of the SMCCC spec. v1.0. This document can be viewed here: https://developer.arm.com/documentation/den0028/a/?lang=en (The `a` in the URL should give you the correct version when you download.). You can query the SMCCC version number on your platform with an SMC call. We did our development on a Zynq UltraScale+ MPSoC, which implements spec v1.0. 

In this version, r4-r7 is just scratch space, so it won't ever be a result. 
Also, since we are using `x0-x7`, we are actually using the 64-bit calling convention.

This does mean that if future processors conform to newer calling convention versions, than we need to create version  specific implementations.

File: libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml [hunk 13]
```diff
@@ -207,4 +217,18 @@
             </error>
         </method>
     </interface>
+    <interface name="seL4_ARM_SMC" manual_name="SMC" cap_description="Capability to allow threads to interact with SMC calls.">
```
[the remark this answers is not in the document]
Comment (Furao) [reply libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml]: Agreed. Updated.

File: libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml (hunk 11 above)
Comment (Furao) [reply libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml]: Updated the description accordingly.

File: src/object/cnode.c (hunk 10 above)
Comment (Furao) [reply src/object/cnode.c:817]: Do you just mean that its weird that the keyword is on its own line?

File: src/object/cnode.c (hunk 9 above)
Comment (Furao) [reply src/object/cnode.c:809]: I suppose it could also be wrapped in an architecture ifdef so that if we ever make an architecture specific `cnode.c` file, then it is obvious that chunk of code should move.

File: src/object/cnode.c (hunk 10 above)
Comment (Indanz) [reply src/object/cnode.c:817]: Yes.

File: src/arch/arm/object/smc.c (hunk 12 above)
Comment (Indanz) [reply src/arch/arm/object/smc.c:32]: We don't want version specific implementations. This is a new multi-platform feature, it should be as future proof as possible, which means following the latest spec. If in the future a platform does return data in registers x4-x7, with the current code it will silently clobber a register and that will be extremely hard to debug if it causes weird problems.

If you do follow the 64-bit calling convention, a lot more registers get clobbered, see table 3-1. The normal procedure call conventions are used, registers `x0-x17`  may be modified. On the upside, you don't have to check bit 30, as you can do 32-bit calls with the 64-bit convention.

With that in mind, maybe it's better to add a `do_smc` function to assembly and call it from here, I think that will get all clobbers and assumptions right automatically.

File: src/object/cnode.c (hunk 10 above)
Comment (kent-mcleod) [reply src/object/cnode.c:817]: I thought that the original weirdness is due to an autostyle edge case. Turns out it's not, as the autostyle accepts it on the same line.. Agree we should change it.

**Comment by kent-mcleod** [comment]
> Yeah, I can pull those in.

Thanks.  I'd be happy to apply any further fixes myself, but the PR branch permissions don't allow modifications by reviewers.

File: src/arch/arm/object/smc.c (hunk 12 above)
Comment (Indanz) [reply src/arch/arm/object/smc.c:32]: I'm fine with limiting it to `x0-x7` for now, as it will take ages before more is practically needed, and when the limit is reached, it's easy to extend. On the other hand, it's probably not much more work to extend it now.

>  I am not entirely sure what an assembly function gets us over the inline assembly we are already doing. The context still needs to be transferred in/out of the registers?

Doing the SMC call would be just a normal function call then. But I forgot that there is a need to transfer the result too, so the advantage would be much less, if any.

Comment (Furao) [reply src/arch/arm/object/smc.c:32]: Playing around with with using all 18 registers revealed to me that inline assembly calls can only have 30 operands, so 18 inputs + 18 outputs doesn't work. Path of least resistance is probably to just modify `r4-r7` to be `r+`, so it is an input and output. 

Or if we really wanted to support all 18 registers, we would need to write a proper assembly function. Thoughts?

Comment (Furao) [reply src/arch/arm/object/smc.c]: I implemented the simpler approach with my latest push.

Comment (Indanz) [reply src/arch/arm/object/smc.c]: Almost there, you also need to add all other modified registers to the clobber list, see Table 3-1.

File: src/arch/arm/object/smc.c [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    register seL4_Word r3 asm("x3") = arg[3];
+    register seL4_Word r4 asm("x4") = arg[4];
+    register seL4_Word r5 asm("x5") = arg[5];
+    register seL4_Word r6 asm("x6") = arg[6];
+    register seL4_Word r7 asm("x7") = arg[7];
+    asm volatile("smc #0\n"
+                 : "+r"(r0), "+r"(r1), "+r"(r2), "+r"(r3),
+                 "+r"(r4), "+r"(r5), "+r"(r6), "+r"(r7)
+                 :
+                 : "memory");
```
Comment (Indanz) [inline src/arch/arm/object/smc.c:34]:
```suggestion
                 :: "x8", "x9", "x10", "x11", "x12", "x13", "x14", "x15", "x16", "x17", "memory");
```

Comment (Furao) [reply src/arch/arm/object/smc.c]: Made the suggested change

File: src/arch/arm/object/smc.c (hunk 12 above)
Comment (Furao) [reply src/arch/arm/object/smc.c]: I appreciate the input. This implementation is much more future-proof now.

**Comment by axel-h** [comment]
I might be good to mention the new `seL4_CapSMC` and [RFC-9](https://sel4.atlassian.net/browse/RFC-9) in the [CHANGES](https://github.com/seL4/seL4/blob/master/CHANGES) file also.

File: src/arch/arm/config.cmake [hunk 15]
```diff
@@ -208,6 +208,17 @@ config_option(
 )
 mark_as_advanced(KernelAArch64SErrorIgnore)
 
+config_option(
+    KernelAllowSMCCalls ALLOW_SMC_CALLS "Allow components to make SMC calls.
+    WARNING: Some platforms require SMC calls to actually be made to boot VMs \
+    correctly, however allowing SMC calls causes a couple of issues. Since seL4 \
+    cannot pre-empt the supervisor the WCET is no longer guaranteed. Also, \
```
Comment (axel-h) [inline src/arch/arm/config.cmake:215]:
```suggestion
    cannot pre-empt the supervisor, the WCET is no longer guaranteed. Also, \
```

Comment (axel-h) [reply src/arch/arm/config.cmake:215]: Also, I think we should use "monitor" (or "secure monitor") instead of "supervisor" here.

File: src/arch/arm/config.cmake [hunk 16]
```diff
@@ -208,6 +208,17 @@ config_option(
 )
 mark_as_advanced(KernelAArch64SErrorIgnore)
 
+config_option(
+    KernelAllowSMCCalls ALLOW_SMC_CALLS "Allow components to make SMC calls.
+    WARNING: Some platforms require SMC calls to actually be made to boot VMs \
```
[the remark this answers is not in the document]
Comment (Furao) [reply src/arch/arm/config.cmake]: Agreed. I removed that line.

File: src/arch/arm/config.cmake (hunk 15 above)
Comment (Furao) [reply src/arch/arm/config.cmake]: Updated both suggestions.

**Comment by Furao** [comment]
> I might be good to mention the new `seL4_CapSMC` and [RFC-9](https://sel4.atlassian.net/browse/RFC-9) in the [CHANGES](https://github.com/seL4/seL4/blob/master/CHANGES) file also.

Added a new commit to update CHANGES file.

File: src/arch/arm/config.cmake (hunk 16 above)
[the remark this answers is not in the document]
Comment (Indanz) [reply src/arch/arm/config.cmake]: Agreed. Other than doing standard PSCI calls, SCM is used for platform specific things. People needing SCM will know whether they need it or not.

**Comment by Furao** [comment]
I just rebased onto the latest and I believe I have addressed all of the open comments. Would be nice to get this through.

**Review by axel-h — APPROVED** [review]
From my side this is fine now. I lack the resources to test this in more detail, tough. 
You could squash the last commit with the CHANGES update into the commit that adds the capability.

**Review by kent-mcleod — APPROVED** [review]

**Comment by kent-mcleod** [comment]
> I just rebased onto the latest and I believe I have addressed all of the open comments. Would be nice to get this through.

I'll wait a couple more days for feedback from other reviewers and otherwise merge.

**Review by Ivan-Velickovic — COMMENTED** [review]
The single comment I had has been addressed, but I’m not qualified to review/approve the actual contents of the patch so I’ll leave that for the others.

**Review by lsf37 — APPROVED** [review]
No further issues from my side. If we had a model of what SMC calls do this does even look fairly verifiable. 

(The preprocess check triggers on the change of `seL4_NumInitialCaps` which should not affect the proofs)

**Comment by lsf37** [comment]
Have remove the proof-test label to preserve AWS instances, because the proofs have passed with this PR now. It looks like the build failed with sel4test because the PR needs to be rebased first to work with the latest libsel4.

File: src/arch/arm/object/smc.c [hunk 17]
```diff
@@ -8,18 +8,17 @@
...
 {
     word_t i;
     seL4_Word arg[NUM_SMC_REGS];
-    word_t *ipcBuffer;
 
     for (i = 0; i < NUM_SMC_REGS; i++) {
         arg[i] = getSyscallArg(i, buffer);
     }
 
-    ipcBuffer = lookupIPCBuffer(true, NODE_STATE(ksCurThread));
```
Comment (Indanz) [inline src/arch/arm/object/smc.c]: I don't think you can safely remove this, however tempting:

 > Require the input length to be >= NUM_SMC_REGS. This ensures that any  accesses to the IPC buffer memory is safe (if the IPC buffer isn't  present then the message length will be truncated to n_msgRegisters).

This misses the corner case where the IPC buffer is mapped read-only.

**Review by Indanz — COMMENTED** [review]
I'd add a back reference to this pull request to the first commit, so people can find this easily when going through the code or git log.

File: src/arch/arm/object/smc.c (hunk 17 above)
Comment (kent-mcleod) [reply src/arch/arm/object/smc.c]: Yea you're right good catch. TIL that IPC buffer caps can be without write rights.  (seems like such an anti-feature)

File: libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml [hunk 18]
```diff
@@ -207,4 +217,20 @@
...
+            </brief>
+            <description>
+                Takes x0-x7 as arguments to an SMC call which are defined as a seL4_ARM_SMCContext
+                struct. The microkernel makes the SMC call and then returns the results as a
+                new seL4_ARM_SMCContext.
+            </description>
+            <param dir="in" name="smc_args" type="seL4_ARM_SMCContext"
+                description="The structure that has the provided arguments."/>
+            <param dir="out" name="smc_response" type="seL4_ARM_SMCContext"
+                description="The structure to capture the responses."/>
```
[the remark this answers is not in the document]
Comment (Furao) [reply libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml:233]: Correct, `inout` is not supported. 

`syscall_stub_gen.py` does the parsing:
```py
assert (param_dir == "in") or (param_dir == "out")
```

File: src/arch/arm/object/smc.c (hunk 17 above)
Comment (Furao) [reply src/arch/arm/object/smc.c]: Added it back in

**Comment by Furao** [comment]
> From my side this is fine now. I lack the resources to test this in more detail, tough. You could squash the last commit with the CHANGES update into the commit that adds the capability.

Squashed that change with the first one.

> I'd add a back reference to this pull request to the first commit, so people can find this easily when going through the code or git log.

Updated the first commit message with the link to this PR.

File: src/arch/arm/object/smc.c [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    word_t smc_func_id = getSyscallArg(0, buffer);
+
+    if (badge != 0 && badge != smc_func_id) {
+        userError("ARMSMCCall: Illegal operation.");
+        current_syscall_error.type = seL4_IllegalOperation;
+        return EXCEPTION_SYSCALL_ERROR;
+    }
+
+    setThreadState(NODE_STATE(ksCurThread), ThreadState_Restart);
+    return invokeSMCCall(buffer, call);
```
Comment (Indanz) [inline src/arch/arm/object/smc.c:91]: `buffer` is not used any more, perhaps move the `lookupIPCbuffer()` call to here and pass that instead of buffer? Or remove the buffer argument, like you had before.

File: libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml (hunk 18 above)
[the remark this answers is not in the document]
Comment (Indanz) [reply libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml:233]: Oh well, so be it, it's not the end of the world and can be changed later if we want.

**Review by Indanz — APPROVED** [review]
Looks good now.

Would be nice if:

https://github.com/seL4/sel4test/pull/85
https://github.com/seL4/sel4test/pull/101

would get merged together, but it's fine if they get merged upstream separately. Could you do a review of Kent's SMC tests?

File: src/arch/arm/object/smc.c [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    seL4_Word arg[NUM_SMC_REGS];
+    word_t *ipcBuffer;
+
+    for (i = 0; i < NUM_SMC_REGS; i++) {
+        arg[i] = getSyscallArg(i, buffer);
+    }
+
+    ipcBuffer = lookupIPCBuffer(true, NODE_STATE(ksCurThread));
+
+    register seL4_Word r0 asm("x0") = arg[0];
```
Comment (axel-h) [inline src/arch/arm/object/smc.c:25]: Maybe this should have a compile time assert that `NUM_SMC_REGS` is at least 8. It's unlikey to ever change, but may catch cases where somebody is experimenting. I'd expect the compiler to issue a out-of-warning in this case automatically, though - but does this also break the build with the current defaults?

File: include/arch/arm/arch/object/structures.h [hunk 21]
```diff
@@ -12,6 +12,16 @@
 
 static inline bool_t CONST Arch_isCapRevocable(cap_t derivedCap, cap_t srcCap)
 {
+#ifdef CONFIG_ALLOW_SMC_CALLS
+    switch (cap_get_capType(derivedCap)) {
+    case cap_smc_cap:
+        return (cap_smc_cap_get_capSMCBadge(derivedCap) !=
+                cap_smc_cap_get_capSMCBadge(srcCap));
+
```
Comment (Indanz) [inline include/arch/arm/arch/object/structures.h:20]: I would remove this empty line, it looks a bit weird.

File: src/arch/arm/object/smc.c (hunk 20 above)
Comment (kent-mcleod) [reply src/arch/arm/object/smc.c:25]: If you want to make that in a separate PR then I'd be happy to accept it.

Comment (axel-h) [reply src/arch/arm/object/smc.c:25]: will do

**Comment by Ivan-Velickovic** [comment]
My understanding is that this PR does not support the SMC64 calling convention, is that correct?

**Comment by Ivan-Velickovic** [comment]
> > My understanding is that this PR does not support the SMC64 calling convention, is that correct?
> 
> No, it's not. You can do both 32-bit and 64-bit SMC calls with this interface. However, it's only implemented for Aarch64, not 32-bit ARM.

The reason I asked is being the SMC specification I was looking at (downloaded from here https://developer.arm.com/documentation/den0028/f?lang=en) says that for SMC64/HVC64 (section 2.7), registers x1-x17 are used for arguments and x0-17 are used for results. This patch takes in x0-x7 as arguments/results.

I *think* I've found a case where I need SMC64, so if I'm happy to add support for the SMC64 calling convention, assuming I've understood everything correctly.

**Comment by Ivan-Velickovic** [comment]
> Technically, we are supporting 32-bit and 64-bit SMC calls but for an earlier version of the spec (v1.0) which didn't include x8-x17 for arguments or results. I looked at adding the rest of the registers, but ran into an issue with inline assembly.

Yep, I just read through the spec changes and saw that in August 2020 it was introduced to use x8-x17 as arguments/results. That makes sense, now. Although, we should document this. I assume then issue B/version 1.1 of the spec was used?

> To support this, you need to do a refactor and write a new function as just assembly.

Yes I ran into the same issue.

**Comment by Ivan-Velickovic** [comment]
Ah GitHub hides all the existing conversations on the calling convention which is why I missed it, my bad.

**Comment by Furao** [comment]
> Ah GitHub hides all the existing conversations on the calling convention which is why I missed it, my bad.

No problem. I knew to look for the comment and it was hard for me to find as well.

> > Technically, we are supporting 32-bit and 64-bit SMC calls but for an earlier version of the spec (v1.0) which didn't include x8-x17 for arguments or results. I looked at adding the rest of the registers, but ran into an issue with inline assembly.
> 
> Yep, I just read through the spec changes and saw that in August 2020 it was introduced to use x8-x17 as arguments/results. That makes sense, now. Although, we should document this. I assume then issue B/version 1.1 of the spec was used?

Yes documentation on the version would be good. Yes, I think 1.1 is technically the latest one we support.
